graphqlapi/src/main/java/io/stargate/graphql/schema/schemafirst/fetchers/dynamic/DynamicFetcher.java

```diff
@@ -231,11 +257,51 @@ private Object toGraphqlUdtValue(UdtValue udtValue) {
       Keyspace keyspace,
       AuthenticationSubject authenticationSubject)
       throws UnauthorizedException {
+    ResultSet resultSet =
+        getResultSet(entity, values, valueNames, dataStore, keyspace, authenticationSubject);
+    if (resultSet == null) return null;
+    Row row = resultSet.one();
+    return mapSingleRow(entity, row);
+  }
+
+  private Map<String, Object> mapSingleRow(EntityMappingModel entity, Row row) {
+    Map<String, Object> singleResult = new HashMap<>();
+    for (FieldMappingModel field : entity.getAllColumns()) {
+      Object cqlValue = row.getObject(field.getCqlName());
+      singleResult.put(
+          field.getGraphqlName(),
+          toGraphqlValue(cqlValue, field.getCqlType(), field.getGraphqlType()));
+    }
+    return singleResult;
+  }
+
+  private ResultSet getResultSet(
+      EntityMappingModel entity,
+      Map<String, Object> values,
+      Optional<List<String>> valueNamesOptional,
+      DataStore dataStore,
+      Keyspace keyspace,
+      AuthenticationSubject authenticationSubject)
+      throws UnauthorizedException {
+    System.out.println(
+        "valueNames: " + valueNamesOptional + "values: " + values + "entity: " + entity);
     List<BuiltCondition> whereConditions = new ArrayList<>();
-    for (int i = 0; i < entity.getPrimaryKey().size(); i++) {
-      FieldMappingModel field = entity.getPrimaryKey().get(i);
+
+    List<FieldMappingModel> primaryKey = entity.getPrimaryKey();
+    if (valueNamesOptional.isPresent()) {
+      List<String> valueNames = valueNamesOptional.get();
+      // use only PKs that are provided in the valueNames
+      // todo should we use pk.getGraphqlName() or pk.getCqlName()?
+      primaryKey =
+          entity.getPrimaryKey().stream()
+              .filter(pk -> valueNames.contains(pk.getCqlName()))
```

Comment:
I am not sure if we should use `pk.getGraphqlName()` or `pk.getCqlName()`?
I think that `pk.getCqlName()` because `entity.getPrimaryKey()` returns cql names.

Reply:
`entity.getPrimaryKey()` returns a list of `FieldMappingModel`, that have both a GraphQL and a CQL name.

This is because we allow users to specify a custom CQL name if they don't like the default mapping. For example given this definition:
```graphql
type User {
  id: ID
  userName: String @cql_column(name: "user_name")
}
```
The first field has the same names. But for the second field graphqlName=userName and cqlName=user_name.

In this method we're trying to bind the parameters of the mutation to the primary key fields in the query. They must have the same types but we allow the names to be different, for example:
```graphql
type Query {
  findUser(id_to_find: ID): User
}
```
`id_to_find` is the first argument, it matches with the first PK field `id` (`entity.getPrimaryKey.get(0)`). So we take the value of that field, convert it to a CQL value, and bind it to the CQL name of that entity field.

I don't follow what you are doing in this loop, I think the code should be very similar to the single-entity variant, except that you stop the iteration earlier because there are less parameters to bind. Maybe you can remember how many parameters the GraphQL query has, and use that as the limit:
```java
    for (int i = 0; i < argumentCount /* instead of entity.getPrimaryKey().size() */; i++) {
      FieldMappingModel field = entity.getPrimaryKey().get(i);
      int finalI = i;
      String inputName = valueNames.map(l -> l.get(finalI)).orElse(field.getGraphqlName());
      Object graphqlValue = values.get(inputName);
      whereConditions.add(
          BuiltCondition.of(
              field.getCqlName(),
              Predicate.EQ,
              toCqlValue(graphqlValue, field.getCqlType(), keyspace)));
    }
```